rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,368 @@
+            /* mask */ 0, /* binary scale */ {DNNL_RUNTIME_F32_VAL});
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(1) | DNNL_ARG_DST,
+            /* mask */ 0, /* eltwise scale */ {DNNL_RUNTIME_F32_VAL});
+
+    ...
+    std::unordered_map<int, memory> args;
+    ...
+    args.insert({DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1 | DNNL_ARG_SCALE, p_op_0_src_1_scale_mem});
+    args.insert({DNNL_ARG_ATTR_MULTIPLE_POST_OP(1) | DNNL_ARG_DST | DNNL_ARG_SCALE, p_op_1_scale_mem});
+    p.execute(s, args);
+    ```
+ - `built-in parameter` in Sum
+   ``` cpp
+   primitive_attr attr;
+   attr.set_scales(DNNL_ARG_MULTIPLE_SRC + 0,
+           /* mask */ 0, {DNNL_RUNTIME_F32_VAL});
+   attr.set_scales(DNNL_ARG_MULTIPLE_SRC + 1,
+           /* mask */ 0, {DNNL_RUNTIME_F32_VAL});
+
+   ...
+   std::unordered_map<int, memory> args;
+   ...
+   args.insert({DNNL_ARG_MULTIPLE_SRC + 0, scale_0_mem});
+   args.insert({DNNL_ARG_MULTIPLE_SRC + 1, scale_1_mem});
+   p.execute(s, args);
+   ```
+
+
+#### Creation-time scaling support will be dropped.
+``` c
+dnnl_status_t DNNL_API dnnl_primitive_attr_get_scales(
+        dnnl_primitive_attr_t attr, int arg, dnnl_dim_t *count, int *mask);
+
+
+dnnl_status_t DNNL_API dnnl_primitive_attr_set_scales(
+        dnnl_primitive_attr_t attr, int arg, dnnl_dim_t count, int mask);
+```
+
+
+#### Simplifying scales configuration
+
+Currently correspondence mask defines the correspondence between the tensor
+dimensions and the scales vector. The set i-th bit indicates that a dedicated
+scaling factor is used for each index along that dimension. The mask value of 0
+implies a common scaling factor for the whole tensor.
+
+The are a few problems:
+1. Setting a mask is complicated. For example, to set per channels scales, mask
+   should be set to `2` (`1 << 1`) for a tensor where channels is a second
+   dimension.
+2. The same scale configuration should be set differently for tensor with
+   different semantics. For example, per channel scales:
+     - rnn data: `1 << 2`, because in rnn dimension order is is `tnc`, so
+       channels is 2nd dimension.
+     - conv data: `1 << 1`, because in conv dimension order is `nchw`, so
+       channels is 1st dimension.
+
+##### Improvement 1
+
+* This option doesn't break API, but simplifies mask usage.
+* oneDNN provides set of pre-defined correspondence masks:
+  ```c
+  // include/oneapi/dnnl/dnnl_types.h
+
+  #define DNNL_MASK_COMMON 0
+  #define DNNL_MASK_PER_DIM_0 1 << 0
+  #define DNNL_MASK_PER_DIM_1 1 << 1
+  #define DNNL_MASK_PER_DIM_2 1 << 2
+  ```
+  ```cpp
+  primitive_attr attr;
+  attr.set_scales(DNNL_ARG_DST, DNNL_MASK_COMMON, scales);
+  ```
+
+##### Improvement 2
+
+* This option extends mask values to express dimensions based on dimension order
+  semantic.
+* oneDNN provides set of pre-defined correspondence masks:
+  ```c
+  // include/oneapi/dnnl/dnnl_types.h
+
+  #define DNNL_MASK_NCHW_PER_CHANNEL 1 << 12
+  #define DNNL_MASK_TNC_PER_CHANNEL 1 << 13
```

Comment:
Such name might bring confusion for batched matmul users, esp. with batch dims > 1.
I would suggest to use `per_channel_nchw` and `per_channel_last` instead.

Could you also clarify why such values and what the implementation would do with it, please?

Reply:
oneDNN supports 12 dimensions, so it is not expected to have regular masks that exceed (1 << 12) - 1. 1 << 12 is the first value that is free for use. Implementation can better check during primitive creation that value is supported. Implementations should treat it equal to per channel output scales. This change is about user's convinience.